File: concepts/10-sorting-algorithms/node/sorting.js
/**
 * Sorting Algorithms in Node.js
 */

/**
 * Bubble Sort - O(n²) worst, O(n) best
 * Stable, in-place
 */
function bubbleSort(arr) {
    const n = arr.length;
    const result = [...arr];

    for (let i = 0; i < n; i++) {
        let swapped = false;
        for (let j = 0; j < n - i - 1; j++) {
            if (result[j] > result[j + 1]) {
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
                swapped = true;
            }
        }
        if (!swapped) break;
    }

    return result;
}

/**
 * Selection Sort - O(n²)
 * Not stable, in-place
 */
function selectionSort(arr) {
    const result = [...arr];
    const n = result.length;

    for (let i = 0; i < n; i++) {
        let minIdx = i;
        for (let j = i + 1; j < n; j++) {
            if (result[j] < result[minIdx]) {
                minIdx = j;
            }
        }
        [result[i], result[minIdx]] = [result[minIdx], result[i]];
    }

    return result;
}

/**
 * Insertion Sort - O(n²) worst, O(n) best
 * Stable, in-place
 */
function insertionSort(arr) {
    const result = [...arr];

    for (let i = 1; i < result.length; i++) {
        const key = result[i];
        let j = i - 1;

        while (j >= 0 && result[j] > key) {
            result[j + 1] = result[j];
            j--;
        }
        result[j + 1] = key;
    }

    return result;
}

/**
 * Merge Sort - O(n log n)
 * Stable, not in-place
 */
function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));

    return merge(left, right);
}

function merge(left, right) {
    const result = [];
    let i = 0, j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }

    return result.concat(left.slice(i)).concat(right.slice(j));
}

/**
 * Quick Sort - O(n log n) average, O(n²) worst
 * Not stable, in-place
 */
function quickSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x < pivot);
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x > pivot);

    return [...quickSort(left), ...middle, ...quickSort(right)];
}

/**
 * Heap Sort - O(n log n)
 * Not stable, in-place
 */
function heapSort(arr) {
    function heapify(arr, n, i) {
        let largest = i;
        const left = 2 * i + 1;
        const right = 2 * i + 2;

        if (left < n && arr[left] > arr[largest]) {
            largest = left;
        }

        if (right < n && arr[right] > arr[largest]) {
            largest = right;
        }

        if (largest !== i) {
            [arr[i], arr[largest]] = [arr[largest], arr[i]];
            heapify(arr, n, largest);
        }
    }

    const result = [...arr];
    const n = result.length;

    // Build max heap
    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
        heapify(result, n, i);
    }

    // Extract elements
    for (let i = n - 1; i > 0; i--) {
        [result[0], result[i]] = [result[i], result[0]];
        heapify(result, i, 0);
    }

    return result;
}

/**
 * Counting Sort - O(n + k) where k is range
 * Stable, not in-place
 */
function countingSort(arr, maxVal = null) {
    if (arr.length === 0) {
        return [];
    }

    if (maxVal === null) {
        maxVal = Math.max(...arr);
    }

    const count = new Array(maxVal + 1).fill(0);

    for (const num of arr) {
        count[num]++;
    }

    const result = [];
    for (let i = 0; i < count.length; i++) {
        for (let j = 0; j < count[i]; j++) {
            result.push(i);
        }
    }

    return result;
}

// Example usage
if (require.main === module) {
    const arr = [64, 34, 25, 12, 22, 11, 90];

    console.log("Original:", arr);
    console.log("Bubble Sort:", bubbleSort(arr));
    console.log("Selection Sort:", selectionSort(arr));
    console.log("Insertion Sort:", insertionSort(arr));
    console.log("Merge Sort:", mergeSort(arr));
    console.log("Quick Sort:", quickSort(arr));
    console.log("Heap Sort:", heapSort(arr));
    console.log("Counting Sort:", countingSort([4, 2, 2, 8, 3, 3, 1]));
}

module.exports = {
    bubbleSort,
    selectionSort,
    insertionSort,
    mergeSort,
    quickSort,
    heapSort,
    countingSort
};

File: concepts/10-sorting-algorithms/python/sorting.py
"""
Sorting Algorithms in Python
"""

def bubble_sort(arr):
    """
    Bubble Sort - O(n²) worst, O(n) best
    Stable, in-place
    """
    n = len(arr)
    arr = arr.copy()
    
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        if not swapped:
            break
    
    return arr


def selection_sort(arr):
    """
    Selection Sort - O(n²)
    Not stable, in-place
    """
    arr = arr.copy()
    n = len(arr)
    
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    
    return arr


def insertion_sort(arr):
    """
    Insertion Sort - O(n²) worst, O(n) best
    Stable, in-place
    """
    arr = arr.copy()
    
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    
    return arr


def merge_sort(arr):
    """
    Merge Sort - O(n log n)
    Stable, not in-place
    """
    if len(arr) <= 1:
        return arr
    
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    
    return merge(left, right)


def merge(left, right):
    """Merge two sorted arrays"""
    result = []
    i = j = 0
    
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    
    result.extend(left[i:])
    result.extend(right[j:])
    return result


def quick_sort(arr):
    """
    Quick Sort - O(n log n) average, O(n²) worst
    Not stable, in-place
    """
    if len(arr) <= 1:
        return arr
    
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    
    return quick_sort(left) + middle + quick_sort(right)


def heap_sort(arr):
    """
    Heap Sort - O(n log n)
    Not stable, in-place
    """
    def heapify(arr, n, i):
        largest = i
        left = 2 * i + 1
        right = 2 * i + 2
        
        if left < n and arr[left] > arr[largest]:
            largest = left
        
        if right < n and arr[right] > arr[largest]:
            largest = right
        
        if largest != i:
            arr[i], arr[largest] = arr[largest], arr[i]
            heapify(arr, n, largest)
    
    arr = arr.copy()
    n = len(arr)
    
    # Build max heap
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    
    # Extract elements
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)
    
    return arr


def counting_sort(arr, max_val=None):
    """
    Counting Sort - O(n + k) where k is range
    Stable, not in-place
    """
    if not arr:
        return []
    
    if max_val is None:
        max_val = max(arr)
    
    count = [0] * (max_val + 1)
    
    for num in arr:
        count[num] += 1
    
    result = []
    for i in range(len(count)):
        result.extend([i] * count[i])
    
    return result


# Example usage
if __name__ == "__main__":
    arr = [64, 34, 25, 12, 22, 11, 90]
    
    print("Original:", arr)
    print("Bubble Sort:", bubble_sort(arr))
    print("Selection Sort:", selection_sort(arr))
    print("Insertion Sort:", insertion_sort(arr))
    print("Merge Sort:", merge_sort(arr))
    print("Quick Sort:", quick_sort(arr))
    print("Heap Sort:", heap_sort(arr))
    print("Counting Sort:", counting_sort([4, 2, 2, 8, 3, 3, 1]))

